src/native_main.nim

```diff
@@ -233,12 +251,22 @@ proc handleMessage(msg: MessageRecv): string =
 
     return $ %* reply # $ converts to string, %* converts to JSON
 
-while true:
-    let strm = newFileStream(stdin)
-    let message = handleMessage(getMessage(strm))
-    let l = pack("@I", message.len)
-
-    write(stdout, l)
-    write(stdout, message) # %* converts the object to JSON
-    flushFile(stdout)
-
+let params = commandLineParams()
+if len(params) > 0 and params[0] == "restart":
+    when defined windows:
+        # need to check this twice so we fail properly on wrong usage of
+        # "restart" argument
+        if len(params) == 4:
+            restartWindowsStep2(firefoxPid = params[1].parseInt(),
+                profiledir = params[2], browsername = params[3])
+            quit(QuitSuccess)
+    quit(QuitFailure)
+else:
+    while true:
+        let strm = newFileStream(stdin)
+        let message = handleMessage(getMessage(strm))
+        let l = pack("@I", message.len)
+
+        write(stdout, l)
+        write(stdout, message) # %* converts the object to JSON
+        flushFile(stdout)
```

Comment:
```suggestion
when defined windows:
    let params = commandLineParams()
    # usage: ./native_main restart [firefoxPid] [profiledir] [browsername]
    if len(params) == 4 and params[0] == "restart":
        restartWindowsStep2(firefoxPid = params[1].parseInt(),
            profiledir = params[2], browsername = params[3])
        quit(QuitSuccess)
    elif len(params) > 0:
        quit(QuitFailure)

while true:
    let strm = newFileStream(stdin)
    let message = handleMessage(getMessage(strm))
    let l = pack("@I", message.len)

    write(stdout, l)
    write(stdout, message) # %* converts the object to JSON
    flushFile(stdout)
```

Untested but perhaps simpler?

Reply:
LGTM. Now that I'm thinking about it, we don't even need two different `quit` calls, since the exit code is guaranteed to not be picked up by the parent process anyway.